src/libtiled/mapobject.cpp

```diff
@@ -209,16 +209,40 @@ void MapObject::flip(FlipDirection direction)
             mCell.setFlippedVertically(!mCell.flippedVertically());
     }
 
-    if (!mPolygon.isEmpty()) {
-        const QPointF center2 = mPolygon.boundingRect().center() * 2;
+    //creating flipmatrix
+    QMatrix flipMatrix;
+    flipMatrix.translate(flipCenter.x(), flipCenter.y());
+    if (direction == FlipHorizontally)
+        flipMatrix.scale(1, -1);
+    else //direction == FlipVertically
+        flipMatrix.scale(-1, 1);
+    flipMatrix.translate(-flipCenter.x(), -flipCenter.y());
 
-        if (direction == FlipHorizontally) {
-            for (int i = 0; i < mPolygon.size(); ++i)
-                mPolygon[i].setX(center2.x() - mPolygon[i].x());
-        } else if (direction == FlipVertically) {
-            for (int i = 0; i < mPolygon.size(); ++i)
-                mPolygon[i].setY(center2.y() - mPolygon[i].y());
-        }
+    //computing position offset
+    QPointF positionOffset;
+    QSizeF objectSize = mPolygon.empty() ? size() : mPolygon.boundingRect().size();
+    if (direction == FlipHorizontally)
+        positionOffset = QPointF(0, objectSize.height());
+    else //direction == FlipVertically
+        positionOffset = QPointF(objectSize.width(), 0);
```

Comment:
Note that the object's origin is unfortunately different depending on whether we're dealing with a shape object or a tile object. Please make sure it works in both cases.